src/Icicle/Avalanche/Check.hs

```diff
@@ -88,7 +88,15 @@ checkStatement frag ctx stmt
                    $ checkExp frag (ctxExp ctx) x
                Let n x' <$> go stmts
 
-        ForeachInts n from to stmts
+        While t n from to stmts
+         -> do from' <- first ProgramErrorExp
+                      $ checkExp frag (ctxExp ctx) from
+               to'   <- first ProgramErrorExp
+                      $ checkExp frag (ctxExp ctx) to
+
+               While t n from' to' <$> go stmts
```

Comment:
hmm. if "n" is supposed to start with some value and be modified, this should probably take an accumulator name, rather than a scalar name.
that way you can just insert "n" into the accumulator map, and inside it can do "Write n".